### PR TITLE
AMMVote

### PR DESCRIPTION
**Pull request by florent-uzio**
# Title

This PR needs to be reviewed after #41 

This PR aims to add the `AMMVote` transaction type with the methods:
- Flatten
- Validate
- The relevant tests for those methods

The coverage for AMMVote is 100%

image

### CODE REVIEW

File: xrpl/transaction/amm_vote.go [hunk 1]
```diff
@@ -1,14 +1,79 @@
...
+//	        "currency" : "TST",
+//	        "issuer" : "rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd"
+//	    },
+//	    "Fee" : "10",
+//	    "Flags" : 2147483648,
+//	    "Sequence" : 8,
+//	    "TradingFee" : 600,
+//	    "TransactionType" : "AMMVote"
+//	}
+//
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_vote.go:31]: Remove comment empty lines

File: xrpl/transaction/amm_vote.go [hunk 2]
```diff
@@ -1,14 +1,79 @@
...
 }
 
 func (*AMMVote) TxType() TxType {
 	return AMMVoteTx
 }
 
-// TODO: Implement flatten
-func (s *AMMVote) Flatten() FlatTransaction {
-	return nil
+func (a *AMMVote) Flatten() FlatTransaction {
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_vote.go:47]: Add comment

File: xrpl/transaction/amm_vote.go [hunk 3]
```diff
@@ -1,14 +1,79 @@
...
 	BaseTx
+	// The definition for one of the assets in the AMM's pool. In JSON, this is an object with currency and issuer fields (omit issuer for XRP).
+	Asset ledger.Asset
+	// The definition for the other asset in the AMM's pool. In JSON, this is an object with currency and issuer fields (omit issuer for XRP).
+	Asset2 ledger.Asset
+	// The proposed fee to vote for, in units of 1/100,000; a value of 1 is equivalent to 0.001%. The maximum value is 1000, indicating a 1% fee.
+	TradingFee uint16
 }
 
 func (*AMMVote) TxType() TxType {
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_vote.go:43]: Add comment

File: xrpl/transaction/amm_vote.go [hunk 4]
```diff
@@ -1,14 +1,79 @@
...
+	// Add AMMDelete-specific fields
+	flattened["TransactionType"] = "AMMVote"
+	flattened["Asset"] = a.Asset.Flatten()
+	flattened["Asset2"] = a.Asset2.Flatten()
+	flattened["TradingFee"] = a.TradingFee
+
+	return flattened
+}
+
+func (a *AMMVote) Validate() (bool, error) {
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_vote.go:60]: Add comment

File: xrpl/transaction/amm_vote.go [hunk 5]
```diff
@@ -1,14 +1,79 @@
...
+	if ok, err := IsAsset(a.Asset); !ok {
+		return false, err
+	}
+
+	if ok, err := IsAsset(a.Asset2); !ok {
+		return false, err
+	}
+
+	if a.TradingFee > 1000 {
+		return false, errors.New("TradingFee must be less than or equal to 1000")
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_vote.go:75]: errors should not be capitalized

**Review by GuillemGarciaDev — CHANGES_REQUESTED** [review]

File: xrpl/transaction/amm_vote.go (hunk 1 above)
Comment (florent-uzio) [reply xrpl/transaction/amm_vote.go:31]: This empty line is added by the Golang formatter in VS Code, it should be fine to leave it like this no?

File: xrpl/transaction/amm_vote.go (hunk 5 above)
Comment (florent-uzio) [reply xrpl/transaction/amm_vote.go]: Good catch

**Review by GuillemGarciaDev — APPROVED** [review]

**Pull request merged**